### PR TITLE
Implement persistence with StatefulSets

### LINKED ISSUE #38: Use ownerReferences with CRD
Using `OwnerReferences` allows us to define relationships between Resources, so that deleting an owner can automatically delete owned resources.

Currently, we use `OwnerReferences` to associate a ConfigMap with a Deployment.

It could be useful to make all Resources we create dependent on the CustomResource, but there might currently be [problems](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#known-issues) with that.

### LINKED ISSUE #44: Use StatefulSets instead of Deployment
Currently we are using [Deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) to deploy our Habitat Service, but since we do not know what are deploying, what type of service that is, it could be anything from a DB to a simple Rails application. We should not just assume our Habitat service would be stateless.

Couple of advantages of [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/):
- Graceful deployment and scaling
- Stable network identity
- Graceful deletion and termination
- Stable, persistent storage

These would be very useful especially if our service is for example a DB.

### LINKED ISSUE #181: RFE: support env and volumeMounts in HabitatSpec
This is a request for two enhancements:

 - [ ] It would be nice to be able to support habitat clusters with persistent disk requirements, such that `volumeMounts` could be defined and taken into account when the Deployment is created in kubernetes. This would enable users to create rings with services that mount a filesystem backed datastore allowing dependent services to query the ring for such things as the db service's ip/port.
 - [x] Additionally, being able to pass supervisor configuration via HAB_SVC_NAME environment variables to the containers would, I believe, be more straightforward and transparent then utilizing base64encoded `toml` implemented as a `Secret`. **[done in #184]**

### PR DESCRIPTION
**Pull request by asymmetric**
This PR

* Changes the type of object we create from `Deployment` to `StatefulSet`
* Adds a `Persistence` key to the `Habitat` type
* Starts splitting `controller.go` into separate modules
* Removes `docs/api.md` and just points users to `types.go`
* ~Adds tests for persistence~

The main issue with using `Deployment`s for persistence was that there was no automatic way to assign a different PVC to each Pod, so we ended up with Pods that shared the same Volume.

See #44 and #195 for more in-depth discussions.

**NOTE**: Tests are included but skipped, as they don't work in our current testing environment (it's not possible to provision `PersistentVolume`s when running minikube with `--vm-driver=none`).

Closes #44.
Closes #38.
Closes #181.

### CODE REVIEW

**Comment by asymmetric** [comment]
@LiliC could you run the deepcopy script against this branch, and see if there's any change?

**Comment by jeremymv2** [comment]
sosweet

File: examples/persisted/habitat.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    # the size of the volume that will be mounted in each Pod
+    size: 1Gi
+    # a StorageClass object with name "standard" must be created beforehand by
+    # the cluster administrator
+    storageClassName: standard
+    # the location under which the volume will be mounted
+    mountPath: /tmp/foobar
+  env:
+    - name: HAB_REDIS
+      value: '{ "protected-mode": "no" }'
```
Comment (LiliC) [inline examples/persisted/habitat.yml]: Please add a note of warning here.

File: pkg/apis/habitat/v1beta1/types.go [hunk 2]
```diff
@@ -51,6 +51,18 @@ type HabitatSpec struct {
 	// The EnvVar type is documented at https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#envvar-v1-core.
 	// Optional.
 	Env []corev1.EnvVar `json:"env,omitempty"`
+	// Optional.
+	Persistence *Persistence `json:"persistence,omitempty"`
+}
+
+type Persistence struct {
```
Comment (LiliC) [inline pkg/apis/habitat/v1beta1/types.go]: Can we call this `Storage` instead? Things make a bit more sense, for example you then have `Size`, `ClassName` as part of `Storage`. WDYT?

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -922,11 +679,20 @@ func habitatKeyFromLabeledResource(r metav1.Object) (string, error) {
 func newConfigMap(ip string, h *habv1beta1.Habitat) *apiv1.ConfigMap {
 	return &apiv1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{
+			// TODO the name should also change to support multiple habitats
```
Comment (LiliC) [inline pkg/controller/controller.go]: Can you explain this `TODO` a bit more?

File: pkg/controller/stateful_sets.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+	d, ok := obj.(*appsv1beta1.StatefulSet)
+	if !ok {
+		level.Error(hc.logger).Log("msg", "Failed to type assert StatefulSet", "obj", obj)
+		return
+	}
+
+	h, err := hc.getHabitatFromLabeledResource(d)
+	if err != nil {
+		// Could not find Habitat, it must have already been removed.
+		level.Debug(hc.logger).Log("msg", "Could not find Habitat for StatefulSet", "name", d.Name)
```
Comment (LiliC) [inline pkg/controller/stateful_sets.go]: Hmm this msg doesn't make much sense to me as a consumer, maybe adding `Could not find Habitat resource for StatefulSet` would be better.

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -183,3 +187,67 @@ func TestHabitatDelete(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestHabitatPersistence(t *testing.T) {
+	// We run minikube in a VM on Travis. In that environment, we cannot create PersistentVolumes.
```
Comment (LiliC) [inline test/e2e/operator_test.go]: Can you explain this a bit, what is the reason behind this?

Also if it doesn't work why do we have this test? I am a bit confused.

File: examples/rbac/rbac.yml [hunk 6]
```diff
@@ -34,7 +34,7 @@ rules:
 - apiGroups:
   - apps
   resources:
-  - deployments
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply examples/rbac/rbac.yml]: No I don't think so, as it's not the operator that's creating `Volume`s, it's the `StatefulSet`.

File: pkg/apis/habitat/v1beta1/zz_generated.deepcopy.go [hunk 7]
```diff
@@ -142,6 +151,22 @@ func (in *HabitatStatus) DeepCopy() *HabitatStatus {
 	return out
 }
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply pkg/apis/habitat/v1beta1/zz_generated.deepcopy.go]: I didn't :) This is your commit actually, from the `Deployment`s PR.

File: pkg/apis/habitat/v1beta1/types.go (hunk 2 above)
Comment (asymmetric) [reply pkg/apis/habitat/v1beta1/types.go]: Not sure that `Storage` by itself makes it clear that this is persistent rather than volatile.

File: pkg/controller/controller.go (hunk 3 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: Oops, that was just for me. Anyway, I think that making the CM name a constant has the drawback that we can only have one. In the future, this could pose problems if we have more than one ring in the same namespace.

File: pkg/controller/stateful_sets.go (hunk 4 above)
Comment (asymmetric) [reply pkg/controller/stateful_sets.go]: Well, it just uses the same format as every other object. We don't say "StatefulSet resource" after all.

File: test/e2e/operator_test.go (hunk 5 above)
Comment (asymmetric) [reply test/e2e/operator_test.go]: The reason is that since we run the operator in containers on Travis, the dynamic volume provisioning that Minikube normally does (when it's on a VM) doesn't work.

I think leaving the test in makes sense because:

* it took time to write it, so we would lose that by deleting it
* we can reactivate it when we move away from testing with minikube (which I think we should)
* it documents that there's a problem and what it is

File: pkg/controller/stateful_sets.go (hunk 4 above)
Comment (LiliC) [reply pkg/controller/stateful_sets.go]: Yes understand, but it can be confusing for a user as the name conveys two things. :) Anyways, more of a nit for clarity.

File: pkg/controller/controller.go (hunk 3 above)
Comment (LiliC) [reply pkg/controller/controller.go]: The problem with is that is how would we handle binding if they are not in one ring. That's why we decided to go with one ring...

File: test/e2e/operator_test.go (hunk 5 above)
Comment (LiliC) [reply test/e2e/operator_test.go]: Personally don't like dead code in master, maybe you could just cherry-pick this and keep it in a branch so the work is not lost.

Comment (asymmetric) [reply test/e2e/operator_test.go]: I think that if it's marked as `skip`, then it's fine. It's common for tests to be temporarily marked as skippable IMO. It would be different if it was business logic, but I think it's fine for tests. I'd rather keep this in unless it's blocking for you.

File: pkg/apis/habitat/v1beta1/types.go (hunk 2 above)
Comment (LiliC) [reply pkg/apis/habitat/v1beta1/types.go]: `PersistantStorage`? :)

Comment (asymmetric) [reply pkg/apis/habitat/v1beta1/types.go]: I'm afraid you're gonna have to re-run the script though :)

File: pkg/controller/controller.go (hunk 3 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: Yes, I didn't mean to say that I thought it through, that's why it was just there for me as a reminder.

**Comment by LiliC** [comment]
@asymmetric I wanted to try this out but can't seem to compile it locally. But it seems to be running fine on travis. :/ Does `make build` work for you?

File: pkg/controller/persistent_volumes.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	_, exists, err := hc.habInformer.GetStore().GetByKey(key)
+	if err != nil {
+		level.Error(hc.logger).Log("msg", "Failed to get key in Store", "obj", key)
+		return
+	} else if !exists {
+		level.Debug(hc.logger).Log("msg", "No matching Habitat found for PVC", "name", pv.Name)
+		return
+	} else {
```
Comment (LiliC) [inline pkg/controller/persistent_volumes.go]: I think we can get rid of this else as we return in other cases anyways.

File: pkg/controller/persistent_volume_claims.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
```
Comment (LiliC) [inline pkg/controller/persistent_volume_claims.go]: It's 2018 :)

File: pkg/controller/persistent_volumes.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
```
Comment (LiliC) [inline pkg/controller/persistent_volumes.go]: It's 2018 :)

File: pkg/controller/stateful_sets.go [hunk 11]
```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
```
Comment (LiliC) [inline pkg/controller/stateful_sets.go]: It's 2018 :)

File: pkg/controller/persistent_volumes.go (hunk 8 above)
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: I think in this case it makes more sense to keep it in in this case, because it's a 3-way situation we're dealing with, and having them together helps (me) map the idea to the code better. 

Usually with 2-branch conditionals that's not necessary, but I find that it helps here.

WDYT?

File: pkg/controller/persistent_volumes.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+	hc.pvInformer = cache.NewSharedIndexInformer(
+		source,
+		&apiv1.PersistentVolume{},
+		resyncPeriod,
+		cache.Indexers{},
+	)
+
+	hc.pvInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: hc.handlePVDelete,
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: I think that at least for now, we can avoid implementing resizes of the storage: it's an optional feature that introduces quite a lot of [complexity](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims). I'd wait for this feature to have stabilized a bit (and for someone to request it) before we figure out how to do it best.

But we should definitely make this clear and notify the user somehow, when they change the `Size` in the manifest, that it's not supported.

**Comment by asymmetric** [comment]
About the `docs/api.md` file: is it useful? I keep forgetting to keep it in sync, and I'd like to know if there's a good reason to do that.

**Comment by LiliC** [comment]
> About the docs/api.md file: is it useful? I keep forgetting to keep it in sync, and I'd like to know if there's a good reason to do that.

I just find the documentation helpful when needing to reference something, but we could also implement a helper script for that? Or just get rid of it, up to you. :)

File: pkg/controller/persistent_volumes.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/go-kit/kit/log/level"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (hc *HabitatController) cachePersistentVolumes() {
+	source := newListWatchFromClientWithLabels(
+		hc.config.KubernetesClientset.CoreV1().RESTClient(),
+		"persistentvolumes",
+		apiv1.NamespaceAll,
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: I guess so, because users can create `Habitat`s in any namespace. Why would this case be different from other resources we watch?

File: pkg/controller/persistent_volumes.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/controller/persistent_volumes.go]: Then IMO I don't see the value of caching PV informer or rather reacting to the changes. Same for the PVC informer. Clean up of volumes is intentionally manual and the rest is already handled by the StatefulSet. The user is already informed, via monitoring, logs, etc., so I feel like we are just duplicating that here. Especially since we are caching ALL the PVC/PV, that can be a huge number in a normal cluster.

File: pkg/controller/controller.go [hunk 14]
```diff
@@ -927,6 +684,14 @@ func newConfigMap(ip string, h *habv1beta1.Habitat) *apiv1.ConfigMap {
 			Labels: map[string]string{
 				habv1beta1.HabitatLabel: "true",
 			},
+			OwnerReferences: []metav1.OwnerReference{
+				metav1.OwnerReference{
+					APIVersion: fmt.Sprintf("%s/%s", habv1beta1.GroupName, habv1beta1.Version),
```
Comment (LiliC) [inline pkg/controller/controller.go]: There is `h.APIVersion` you can use.

File: pkg/controller/controller.go [hunk 15]
```diff
@@ -927,6 +684,14 @@ func newConfigMap(ip string, h *habv1beta1.Habitat) *apiv1.ConfigMap {
...
+					Kind:       "Habitat",
```
Comment (LiliC) [inline pkg/controller/controller.go]: Same here there is `h.Kind` you can use.

File: pkg/controller/persistent_volumes.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: We are only caching ones with label `habitat: true`. Also, the point of caching is that they could be  removed by anything/anybody else, and we need to be able to notify the user (and in the future, update the `Status`).

According to the Controller Bible, verse 7:

> There are other actors in the system. Just because you haven't changed an object doesn't mean that somebody else hasn't.

File: pkg/controller/stateful_sets.go [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+		// Pass --bind flag.
+		bindArg := fmt.Sprintf("%s:%s.%s", bind.Name, bind.Service, bind.Group)
+		habArgs = append(habArgs,
+			"--bind", bindArg)
+	}
+
+	base := &appsv1beta1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: h.Name,
+		},
```
Comment (LiliC) [inline pkg/controller/stateful_sets.go]: Add a `OwnerReference` to the `StatefulSet` as well, like above in the `ConfigMap`.

File: pkg/controller/controller.go [hunk 17]
```diff
@@ -927,6 +684,14 @@ func newConfigMap(ip string, h *habv1beta1.Habitat) *apiv1.ConfigMap {
...
+					Name:       h.Name,
+					UID:        h.UID,
+				},
```
Comment (LiliC) [inline pkg/controller/controller.go]: Also I noticed we do not say anything about `BlockOwnerDeletion`, I would add this here. It is very useful IMO.

File: pkg/controller/stateful_sets.go (hunk 16 above)
Comment (asymmetric) [reply pkg/controller/stateful_sets.go]: We currently do that [here](https://github.com/habitat-sh/habitat-operator/blob/aaca35c5dee80138e793f465ebbd1ddc95d2b10f/pkg/controller/controller.go#L485), but I agree we could simply delegate this to Kubernetes.

File: pkg/controller/controller.go (hunk 17 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: > if the owner has the "foregroundDeletion" finalizer

Is this our case?

It's not entirely clear for me if any defaults are defined for the owning resource (`Habitat`). Otherwise, it seems to me that `propagationPolicy` could be specified by a user in their `DELETE` request. Is this something we should be worried about?

File: pkg/controller/persistent_volumes.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/controller/persistent_volumes.go]: > and we need to be able to notify the user

But that is already done by its respective controller.

Agree that we should react, but we are not doing anything apart from logging. So I don't see the point of it, IMHO. :)

File: pkg/controller/stateful_sets.go (hunk 16 above)
Comment (asymmetric) [reply pkg/controller/stateful_sets.go]: Done.

File: pkg/controller/controller.go (hunk 17 above)
Comment (LiliC) [reply pkg/controller/controller.go]: Not sure if it's something to worry about, just thought it was nice to have.

File: pkg/controller/persistent_volumes.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: We will update the `Status` in #200 :)

File: pkg/controller/persistent_volumes.go [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+	// Find out if there's a Habitat object around
+	// If so, we have a problem.
+	key, err := cache.MetaNamespaceKeyFunc(pv.Spec.ClaimRef)
+	if err != nil {
+		level.Error(hc.logger).Log("msg", "Failed to get key", "obj", pv)
+		return
+	}
+
+	_, exists, err := hc.habInformer.GetStore().GetByKey(key)
```
Comment (krnowak) [inline pkg/controller/persistent_volumes.go]: That part of code looks like a copy pasta of the `findHabForPVC` function. Maybe just call it here?

File: pkg/controller/stateful_sets.go [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+
+		secretVolumeMount := &apiv1.VolumeMount{
+			Name: userConfigFilename,
+			// The Habitat supervisor creates a directory for each service under /hab/svc/<servicename>.
+			// We need to place the user.toml file in there in order for it to be detected.
+			MountPath: fmt.Sprintf("/hab/user/%s/config", h.Spec.Service.Name),
+			ReadOnly:  false,
+		}
+
+		base.Spec.Template.Spec.Containers[0].VolumeMounts = append(base.Spec.Template.Spec.Containers[0].VolumeMounts, *secretVolumeMount)
```
Comment (krnowak) [inline pkg/controller/stateful_sets.go]: You could probably do `spec := &base.Spec.Template.Spec` and then `spec.Containers[0].VolumeMounts = …` to drop a bit of a repeating the same thing.

Comment (krnowak) [reply pkg/controller/stateful_sets.go]: Or even move this `spec := …` outside this block, so it could be also used for mounting the persistent volume or the ring secret volume.

File: pkg/controller/controller.go [hunk 20]
```diff
@@ -71,14 +70,18 @@ type HabitatController struct {
...
+	pvcInformer cache.SharedIndexInformer
 
 	// cache.InformerSynced returns true if the store has been synced at least once.
-	habInformerSynced    cache.InformerSynced
-	deployInformerSynced cache.InformerSynced
-	cmInformerSynced     cache.InformerSynced
+	habInformerSynced cache.InformerSynced
+	stsInformerSynced cache.InformerSynced
+	cmInformerSynced  cache.InformerSynced
+	pvInformerSynced  cache.InformerSynced
```
Comment (krnowak) [inline pkg/controller/controller.go]: What are `pvInformerSynced` and `pvcInformerSynced` for? They seem to be unused. Did you forget to use them in `WaitForCacheSync` call?

File: pkg/controller/persistent_volumes.go (hunk 18 above)
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: Actually, I think there are 2 problems with my approach here:

* We are not telling users that they `PersistentVolume`s they create (if they're going with static provisioning) need to have the `Habitat` label
* We find the Habitat by name, which seems like a weak approach. Maybe a label would be better? Considering that the PV might be created beforehand, it seems weird to force the user to know already what the name should be. A label, OTOH, can be added later, and seems a better approach than asking the user to rename the PV.

File: pkg/controller/persistent_volumes.go (hunk 8 above)
Comment (asymmetric) [reply pkg/controller/persistent_volumes.go]: I ended up agreeing with you :)

File: examples/rbac/rbac.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (asymmetric) [reply examples/rbac/rbac.yml]: You were right btw, we did have to: the ListWatcher running in the operator means we need to assign `list` and `watch` permissions.

**Comment by asymmetric** [comment]
PTALA, I reverted all commits that dealt with watching PVs/PVCs. I haven't removed them from history because I thought it would be nice to have them for when we'll re-implement this.

Let me know what you think.

File: pkg/controller/controller.go [hunk 21]
```diff
@@ -347,15 +274,15 @@ func (hc *HabitatController) enqueueCM(obj interface{}) {
 }
 
 func (hc *HabitatController) handleCMAdd(obj interface{}) {
-	hc.enqueueCM(obj)
+	hc.handleCM(obj)
```
Comment (LiliC) [inline pkg/controller/controller.go]: Not sure if this is in the scope of this PR or not, but since you are redoing the logic of resources being created or not, I open an issue about CM not being deleted when Habitat resource is deleted, so maybe it's something worth looking into. It can be outside of this PR as well, up to you...

File: pkg/controller/stateful_sets.go [hunk 22]
```diff
@@ -0,0 +1,335 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log/level"
+	habv1beta1 "github.com/habitat-sh/habitat-operator/pkg/apis/habitat/v1beta1"
```
Comment (LiliC) [inline pkg/controller/stateful_sets.go]: My personal preference is to have the these imports separated with a line break from the outside imports.

File: pkg/controller/stateful_sets.go [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const persistentVolumeName = "persistent"
```
Comment (LiliC) [inline pkg/controller/stateful_sets.go]: Why is this name hardcoded?

File: test/e2e/framework/helpers.go [hunk 24]
```diff
@@ -223,7 +224,7 @@ func ConvertSecret(pathToYaml string) (*v1.Secret, error) {
 	return &s, nil
 }
 
-func convertToK8sResource(pathToYaml string, into interface{}) error {
+func convertToK8sResource(pathToYaml string, into runtime.Object) error {
```
Comment (LiliC) [inline test/e2e/framework/helpers.go]: cool! TIL

**Comment by LiliC** [comment]
BTW since this is a huge breaking change IMO we really need to bump up the version here of the API, so to `v2beta1` or something like that. As that we don't get problems if this operator is deployed into a cluster where Habitat resources, e.g. Deployments already exist as that cannot be handled by this operator.

File: examples/rbac/rbac.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (asymmetric) [reply examples/rbac/rbac.yml]: Exactly.

File: pkg/controller/stateful_sets.go (hunk 22 above)
Comment (asymmetric) [reply pkg/controller/stateful_sets.go]: Because it never changes ;)

It's a bit like the name for the CM volume. We hardcode it, and since we are good programmers, we move it to a constant ;)

Basically, there's no reason to expose this name and make it configurable I think.

File: pkg/controller/stateful_sets.go (hunk 23 above)
Comment (asymmetric) [reply pkg/controller/stateful_sets.go]: Because it never changes ;)

It's a bit like the name for the CM volume. We hardcode it, and since we are good programmers, we move it to a constant ;)

Basically, there's no reason to expose this name and make it configurable I think.

File: pkg/controller/stateful_sets.go (hunk 22 above)
Comment (asymmetric) [reply pkg/controller/stateful_sets.go]: Your wish is my command.

File: pkg/controller/controller.go (hunk 21 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: :+1:. I'll tackle it in another PR.

**Comment by asymmetric** [comment]
We agreed offline to keep the current `Deployment`-based behavior in `v1beta1`, alongside the new `StatetfulSet`-based one on `v1beta2`. This will be done in a folloging PR.

So please go ahead and do your final reviews :)

File: pkg/apis/habitat/v1beta1/types.go [hunk 25]
```diff
@@ -51,6 +51,20 @@ type HabitatSpec struct {
...
+// PersistentStorage contains the details of the persistent storage that the
+// cluster should provision.
+type PersistentStorage struct {
+	// Size is the volume's size.
+	// It uses the same format as Kubernetes' size fields, e.g. 10Gi
+	Size string `json:"size"`
+	// MountPath is the path at which the PersistentVolume will be mounted.
+	MountPath string `json:"mountPath"`
+	// StorageClassName is the name of the StorageClass that the StatefulSet will request.
+	StorageClassName string `json:"storageClassName"`
```
Comment (LiliC) [inline pkg/apis/habitat/v1beta1/types.go]: Nit: Can we name this just `ClassName` as it's under the `PersistantStorage` anyways. WDYT?

File: pkg/apis/habitat/v1beta1/types.go [hunk 26]
```diff
@@ -51,6 +51,20 @@ type HabitatSpec struct {
...
+	// Optional.
+	PersistentStorage *PersistentStorage `json:"persistentStorage,omitempty"`
+}
+
+// PersistentStorage contains the details of the persistent storage that the
+// cluster should provision.
+type PersistentStorage struct {
+	// Size is the volume's size.
+	// It uses the same format as Kubernetes' size fields, e.g. 10Gi
+	Size string `json:"size"`
```
Comment (LiliC) [inline pkg/apis/habitat/v1beta1/types.go]: You are missing the `omitempty`, or are these are required? Then we should document that as we don't have the api docs.md anymore. :)

File: pkg/controller/controller.go [hunk 27]
```diff
@@ -927,6 +652,14 @@ func newConfigMap(ip string, h *habv1beta1.Habitat) *apiv1.ConfigMap {
 			Labels: map[string]string{
 				habv1beta1.HabitatLabel: "true",
 			},
+			OwnerReferences: []metav1.OwnerReference{
+				metav1.OwnerReference{
+					APIVersion: habv1beta1.SchemeGroupVersion.String(),
+					Kind:       habv1beta1.HabitatKind,
```
Comment (LiliC) [inline pkg/controller/controller.go]: What do we use this OwnerReference for now?

File: examples/persisted/README.md [hunk 28]
```diff
@@ -0,0 +1,37 @@
+# Persistent Storage example
```
Comment (LiliC) [inline examples/persisted/README.md]: So the whole point of persistence is that you can mount already existing volumes, can you maybe write up some steps of how a user can do that.

File: examples/persisted/README.md [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+
+Habitat objects are translated by the operator into `StatefulSet` objects, which
```
Comment (LiliC) [inline examples/persisted/README.md]: Nit: Maybe add a note that the user needs to know about basic understanding of persistent volumes in K8s and link to it?

File: pkg/controller/stateful_sets.go [hunk 30]
```diff
@@ -0,0 +1,336 @@
...
+
+func (hc *HabitatController) newStatefulSet(h *habv1beta1.Habitat) (*appsv1beta1.StatefulSet, error) {
```
Comment (LiliC) [inline pkg/controller/stateful_sets.go]: You know now that you extracted this you can actually write a unit test for this. WDYT?

File: pkg/controller/controller.go [hunk 31]
```diff
@@ -802,15 +526,16 @@ func (hc *HabitatController) processNextItem() bool {
 
 // conform is where the reconciliation takes place.
 // It is invoked when any of the following resources get created, updated or deleted:
-// Habitat, Pod, Deployment, ConfigMap.
+// Habitat, Pod, StatefulSet, ConfigMap.
 func (hc *HabitatController) conform(key string) error {
 	obj, exists, err := hc.habInformer.GetStore().GetByKey(key)
```
Comment (LiliC) [inline pkg/controller/controller.go]: Should we not handle conforming the `PersistentVolumeClaim` as well? Or is there nothing that needs to be done, when for example Habitat resource gets deleted or updated?

File: pkg/apis/habitat/v1beta1/types.go (hunk 25 above)
Comment (asymmetric) [reply pkg/apis/habitat/v1beta1/types.go]: I gave it the same name it has in the `PersistentVolume` resource, because it [maps 1:1 with that field](
https://github.com/kinvolk/habitat-operator/tree/efe0e8fe74f3a93f550e606b9abad2bf81cd31b6/pkg/controller/stateful_sets.go#L206). My hope is that it will make it easier for people to understand how things tie together.

File: pkg/apis/habitat/v1beta1/types.go (hunk 26 above)
Comment (asymmetric) [reply pkg/apis/habitat/v1beta1/types.go]: What do you mean? Right now we only explicitly document that a field is optional, otherwise it's mandatory.

File: pkg/controller/controller.go (hunk 27 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: In theory so that the CM is deleted with the Habitat, but apparently that's [not working](#213)?

File: examples/persisted/README.md (hunk 28 above)
Comment (asymmetric) [reply examples/persisted/README.md]: I explicitly avoided to do that because it's a big topic and I would only make a poorer job at explaining it than the official docs do, which I linked. I anyway added some commits after your comment which should hopefully make things clearer.

BTW, it's not necessary that the PV has been created beforehand, if the cluster supports dynamic provisioning.

TL;DR: IMHO explaining how persistence works in Kubernetes is beside the scope of this README.

File: examples/persisted/README.md (hunk 29 above)
Comment (asymmetric) [reply examples/persisted/README.md]: OK, I'll add a link to [this](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)

File: pkg/controller/controller.go (hunk 27 above)
Comment (LiliC) [reply pkg/controller/controller.go]: That doesn't link to anything for me? 🤔

File: pkg/controller/controller.go (hunk 31 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: As mentioned in the [example README](https://github.com/habitat-sh/habitat-operator/blob/64e7f10e39792073294f891d4a70ebdaf3a1a51b/examples/persisted/README.md), we do nothing when they get deleted, because we want to prioritize data safety (which is also the default behavior with STS).

Do you think we should do something?

File: pkg/apis/habitat/v1beta1/types.go (hunk 25 above)
Comment (LiliC) [reply pkg/apis/habitat/v1beta1/types.go]: Cool, that makes sense!

File: pkg/apis/habitat/v1beta1/types.go (hunk 26 above)
Comment (LiliC) [reply pkg/apis/habitat/v1beta1/types.go]: Ok, just before in the docs.md file it was more clear at least to me, what is required. But if it makes sense to you 👍

File: examples/rbac/minikube.yml [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+  - kind: ServiceAccount
+    name: default
+    namespace: kube-system
+  - kind: User
+    name: kube-proxy
+roleRef:
+  kind: ClusterRole
+  name: cluster-writer
+  apiGroup: rbac.authorization.k8s.io
+
```
Comment (LiliC) [inline examples/rbac/minikube.yml]: Line too much?

File: examples/rbac/minikube.yml [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: cluster-writer
+rules:
+  - apiGroups: ["*"]
+    resources: ["*"]
+    verbs: ["*"]
+  - nonResourceURLs: ["*"]
+    verbs: ["*"]
+
```
Comment (LiliC) [inline examples/rbac/minikube.yml]: One line too much. :)

File: examples/rbac/minikube.yml [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+    namespace: kube-system
+  - kind: User
+    name: kube-proxy
+roleRef:
+  kind: ClusterRole
+  name: cluster-writer
+  apiGroup: rbac.authorization.k8s.io
+
+---
+
```
Comment (LiliC) [inline examples/rbac/minikube.yml]: Remove line? :)

File: examples/persisted/README.md [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+* familiarize yourself with how persistent storage [works in
+Kubernetes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
+* create a
+[`StorageClass`](https://kubernetes.io/docs/concepts/storage/storage-classes/)
+object with name `example-sc`
+* add the `spec.persistentStorage` key to the Habitat object's manifest
+* specify the `name` of the aforementioned `StorageClass` object under
+`spec.persistence.storageClassName` in the Habitat object's manifest
+
+An example `StorageClass` for clusters running on minikube is provided in
```
Comment (LiliC) [inline examples/persisted/README.md]: Maybe note it doesn't work if it's vm-driver=none?

File: pkg/controller/controller.go (hunk 27 above)
Comment (asymmetric) [reply pkg/controller/controller.go]: Fixed.

File: examples/rbac/minikube.yml (hunk 33 above)
Comment (asymmetric) [reply examples/rbac/minikube.yml]: It's required/recommended to have a newline at the end of a file. My editor does it automatically.

Git complains if you don't have it.

File: examples/rbac/minikube.yml (hunk 34 above)
Comment (asymmetric) [reply examples/rbac/minikube.yml]: I actually think that's not even needed, not sure why it's there in the first place. There's no `sd-build` `ServiceAccount` on my minikube cluster.

File: examples/persisted/README.md (hunk 35 above)
Comment (asymmetric) [reply examples/persisted/README.md]: I don't think anyone's gonna try it with that flag, considering it's not even documented ;)

File: examples/rbac/minikube.yml (hunk 33 above)
Comment (asymmetric) [reply examples/rbac/minikube.yml]: Except that's not the end :rofl:

File: examples/persisted/README.md (hunk 35 above)
Comment (LiliC) [reply examples/persisted/README.md]: Think it is now. :D

File: examples/rbac/minikube.yml (hunk 34 above)
Comment (asymmetric) [reply examples/rbac/minikube.yml]: Wrong position :joy: My comment referred to the bottom `ClusterRoleBinding`.

File: pkg/apis/habitat/v1beta1/types.go (hunk 26 above)
Comment (asymmetric) [reply pkg/apis/habitat/v1beta1/types.go]: AFAICT that's also how its' done in upstream.
